packages/server/lib/remote_states.ts

```diff
@@ -0,0 +1,154 @@
+
+    return this.remoteStates.get(cors.getOriginPolicy(url))
+  }
+
+  isInOriginStack (url: string): boolean {
+    return this.originStack.includes(cors.getOriginPolicy(url))
+  }
+
+  isSecondaryOrigin (url: string): boolean {
+    // start at 1 to exclude the primary origin
+    return this.originStack.indexOf(cors.getOriginPolicy(url), 1) !== -1
+  }
+
+  isPrimaryOrigin (url: string): boolean {
+    return this.originStack[0] === cors.getOriginPolicy(url)
+  }
+
+  addOrigin (originPolicy) {
+    this.originStack.push(originPolicy)
+
+    debug('Added origin: ', originPolicy)
+  }
+
+  removeCurrentOrigin () {
+    const originPolicy = this.originStack.pop()
+
+    debug('Removed current origin: ', originPolicy)
+  }
+
+  reset () {
+    debug('Resetting remote state')
+
+    const stateArray = Array.from(this.remoteStates.entries())
+
+    this.remoteStates = new Map([stateArray[0]])
+
+    this.originStack = [stateArray[0][0]]
+  }
+
+  current (): RemoteState {
+    const state = this.remoteStates.get(this.originStack[this.originStack.length - 1]) as RemoteState
+
+    debug('Getting current remote state: %o', state)
+
+    return _.cloneDeep(state)
+  }
+
+  set (urlOrState: string | RemoteState, options: { auth?: {}, isMultiDomain?: boolean } = {}): RemoteState {
+    let state
+
+    if (_.isString(urlOrState)) {
+      const remoteOrigin = origin(urlOrState)
+      const remoteProps = cors.parseUrlIntoDomainTldPort(remoteOrigin)
+
+      if ((urlOrState === '<root>') || !fullyQualifiedRe.test(urlOrState)) {
+        if (!this.config) {
+          this.config = this.configure()
+        }
+
+        state = {
+          auth: options.auth,
+          origin: `http://${DEFAULT_DOMAIN_NAME}:${this.config.serverPort}`,
+          strategy: 'file',
+          fileServer: _.compact([`http://${DEFAULT_DOMAIN_NAME}`, this.config.fileServerPort]).join(':'),
+          domainName: DEFAULT_DOMAIN_NAME,
+          props: null,
+        }
+      } else {
+        state = {
+          auth: options.auth,
+          origin: remoteOrigin,
+          strategy: 'http',
+          fileServer: null,
+          domainName: cors.getDomainNameFromParsedHost(remoteProps),
+          props: remoteProps,
+        } as RemoteState
+      }
+    } else {
+      state = urlOrState
+    }
+
+    const remoteOriginPolicy = cors.getOriginPolicy(state.origin)
+
+    if (options.isMultiDomain) {
+      this.remoteStates.set(remoteOriginPolicy, state)
+    } else {
+      // convert map to array
+      const stateArray = Array.from(this.remoteStates.entries())
+
+      // set the primary remote state and convert back to map
+      stateArray[0] = [remoteOriginPolicy, state]
+      this.remoteStates = new Map(stateArray)
+
+      // automatically update the primary origin stack
+      this.originStack[0] = remoteOriginPolicy
+    }
+
+    debug('Setting remote state %o for %s', state, remoteOriginPolicy)
+
+    return _.cloneDeep(state)
```

Comment:
I wonder if we want to return `undefined` here and encourage devs to call `set` and then `current`/`get` immediately after if state is needed after setting to prevent any drift between how state is returned.

Reply:
In some use cases the dev may not entirely know the origin since it could be a local reference. I decided to call `.get` from here to ensure there is no drift. [4f347a8](https://github.com/cypress-io/cypress/pull/20752/commits/4f347a811e23fbadb1456cce535a208a96108146)